python_modules/dagster/dagster/_core/definitions/assets.py

```diff
@@ -122,9 +126,9 @@ class AssetsDefinition(ResourceAddable, RequiresResources, IHasInternalInit):
     def __init__(
         self,
         *,
-        keys_by_input_name: Mapping[str, AssetKey],
-        keys_by_output_name: Mapping[str, AssetKey],
-        node_def: NodeDefinition,
+        keys_by_input_name: Mapping[str, AssetKey] = {},
+        keys_by_output_name: Mapping[str, AssetKey] = {},
```

Comment:
I'm always a little disturbed by using mutable objects as default param names (they're not created fresh every execution). Can we use `None` instead and generate `{}` in the function body?

Reply:
I do not like this pattern. We cannot trust the type system enough and someone will subvert it at some point and introduce a really difficult bug to trackdown. This doesn't pass the cost/benefit ratio me. I feel quite strongly that we should continue to standarize on `None`, even for things typed as `Mapping`.

Reply:
My reservations about the `None` direction are:
- It leaves users to wonder whether `None` and `{}` are treated differently.
- It's an irreversible API expansion.

Thoughts on this?
```python
from types import MappingProxyType

EMPTY_FROZEN_DICT = MappingProxyType({})

...

    keys_by_output_name: Mapping[str, AssetKey] = EMPTY_FROZEN_DICT
```

Reply:
As a user if I saw anything other than `None` I would assume it is a bug and that the team is not following established Python best practices. We should not be innovating here. The standards exist for good reason, as Python's behavior in this domain is crazy.

We should not try to change policy coupled with a change that has time urgency. The conversations need to be decoupled.

I understand that this decision is irreversible, but we already default to `None` in our public APIs, so I would argue that is itself a species breaking change, since it changes expectations.

Reply:
Btw I think accepting `None` here might actually be the right call here ultimately, but sounds like there's some urgency to merge this PR to help with a perf issue, and don't want to rush into that.

Reply:
> I understand that this decision is irreversible, but we already default to None in our public APIs, so I would argue that is itself a species breaking change, since it changes expectations

Definitely agree that it's a change with potential downside. I just hate widening the API when it doesn't feel absolutely necessary.

That said, not something I feel super strongly about here - updated the PR.